lib/SIL/Utils/BasicBlockUtils.cpp

```diff
@@ -372,3 +372,146 @@ void DeadEndBlocks::compute() {
+  // Until the worklist is empty...
+  while (!worklist.empty()) {
+    // Grab the next block to visit.
+    SILBasicBlock *block = worklist.pop_back_val();
+
+    // Since the block is on our worklist, we know already that it is not a
+    // block with lifetime ending uses, due to the invariants of our loop.
+
+    // First remove BB from the SuccessorBlocksThatMustBeVisited list. This
+    // ensures that when the algorithm terminates, we know that BB was not the
+    // beginning of a non-covered path to the exit.
+    successorBlocksThatMustBeVisited.remove(block);
+
+    // Ok, now we know that we do not have an overconsume. If this block does
+    // not end in a no return function, we need to update our state for our
+    // successors to make sure by the end of the traversal we visit them.
+    //
+    // We must consider such no-return blocks since we may be running during
+    // SILGen before NoReturn folding has run.
+    for (auto *succBlock : block->getSuccessorBlocks()) {
+      // If we already visited the successor, there is nothing to do since we
+      // already visited the successor.
+      if (visitedBlocks.count(succBlock))
+        continue;
+
+      // Then check if the successor is a transitively unreachable block. In
+      // such a case, we ignore it since we are going to leak along that path.
+      if (deBlocks.isDeadEnd(succBlock))
+        continue;
+
+      // Otherwise, add the successor to our SuccessorBlocksThatMustBeVisited
+      // set to ensure that we assert if we do not visit it by the end of the
+      // algorithm.
+      successorBlocksThatMustBeVisited.insert(succBlock);
+    }
+
+    // If we are at the dominating block of our walk, continue. There is nothing
+    // further to do since we do not want to visit the predecessors of our
+    // dominating block. On the other hand, we do want to add its successors to
+    // the successorBlocksThatMustBeVisited set.
+    if (block == definingBlock)
+      continue;
+
+    // Then for each predecessor of this block:
+    //
+    // 1. If we have visited the predecessor already, then it is not a block
+    // with lifetime ending uses. If it is a block with uses, then we have a
+    // double release... so assert. If not, we continue.
+    //
+    // 2. We add the predecessor to the worklist if we have not visited it yet.
+    for (auto *predBlock : block->getPredecessorBlocks()) {
+      // Check if we have an over consume.
+      if (predBlock == reachableBlock) {
+        for (auto *succ : reachableBlock->getSuccessorBlocks()) {
+          if (!visitedBlocks.count(succ))
+            successorBlocksThatMustBeVisited.insert(succ);
+        }
+        foundLoop = true;
+      }
+
+      if (visitedBlocks.count(predBlock)) {
+        continue;
+      }
+
+      visitedBlocks.insert(predBlock);
+      worklist.push_back(predBlock);
+    }
+  }
+}
+
+bool swift::completePostDominatingSet(
+    SILBasicBlock *definingBlock, SILBasicBlock *reachableBlock,
```

Comment:
It seems odd for a utility to expect a single `reachableBlock`. In most cases don't we have a set of reachable blocks?